File: src/main/java/com/bazbatlabs/bricksmash/models/Brick.java
package com.bazbatlabs.bricksmash.models;

public final class Brick implements Collidable {

    private static final int BREAKING_DURATION = 19;

    private final WorldEvents events;

    private final Rect bounds;
    private State state;
    private int stateCounter;
    private int durability;

    public Brick(Rect bounds, Type type, WorldEvents events) {

        this.events = events;

        this.bounds = bounds;
        this.state = State.NORMAL;
        this.stateCounter = 0;

        if (type == Type.TOUGH) {
            this.durability = 2;
        } else {
            this.durability = 1;
        }
    }

    public boolean isActive() { return state == State.NORMAL; }
    public boolean isDestroyed() { return state == State.DESTROYED; }
    public boolean isBreaking() { return state == State.BREAKING; }
    public int stateCounter() { return stateCounter; }

    public Rect bounds() { return bounds; }

    public void hit() {
        events.enqueue(WorldEvents.Event.BRICK_HIT);

        if (state == State.NORMAL) {
            durability--;

            if (durability <= 0) {
                state = State.BREAKING;
            }
        }
    }

    public void update() {
        if (state == State.BREAKING) {
            stateCounter++;

            if (stateCounter > BREAKING_DURATION) {
                state = State.DESTROYED;
            }
        }
    }

    public Vec2 deflect(Vec2 vel, Vec2 collision, Axis axis) {
        switch (axis) {
        case X: return new Vec2(-vel.x(), vel.y());
        case Y: return new Vec2(vel.x(), -vel.y());
        default: return Vec2.Zero();
        }
    }

    private enum State {
        NORMAL, BREAKING, DESTROYED
    }

    public enum Type {
        NORMAL, TOUGH
    }
}

File: src/main/scala/models/Brick.scala
package com.bazbatlabs.bricksmash.models

class Brick(val bounds: Rect, val brickType: Brick.Type.Value, val events: WorldEvents) extends Collidable {

  val BreakingDuration = 19

  val foo = Brick.Type.Normal

  var state = Brick.State.Normal
  var stateCounter = 0

  var durability = if (brickType == Brick.Type.Tough) 2 else 1

  def isActive = state == Brick.State.Normal
  def isDestroyed = state == Brick.State.Destroyed
  def isBreaking = state == Brick.State.Breaking

  def hit() {
    events.enqueue(Event.BrickHit)

    if (state == Brick.State.Normal) {
      durability -= 1

      if (durability <= 0) {
        state = Brick.State.Breaking
      }
    }
  }

  def update() {
    if (state == Brick.State.Breaking) {
      stateCounter += 1

      if (stateCounter > BreakingDuration) {
        state = Brick.State.Destroyed
      }
    }
  }

  def deflect(vel: Vec2, collision: Vec2, axis: Axis.Value) =
    if (axis == Axis.X) Vec2(-vel.x, vel.y) else Vec2(vel.x, -vel.y)
}

object Brick {
  object Type extends Enumeration {
    type Type = Value
    val Normal = Value("Tough")
    val Tough = Value("Tough")
  }


  object State extends Enumeration {
    type State = Value
    val Normal = Value("Normal")
    val Breaking = Value("Breaking")
    val Destroyed = Value("Destroyed")
  }
}
